File: algorithms/sort.py
import unittest
import abc

class SortTest(unittest.TestCase, abc.ABC):
    @abc.abstractmethod
    def sort(self, nums):
        pass

    def test_1(self):
        _input = []
        self.sort(_input)
        self.assertSequenceEqual(_input, sorted(_input))

    def test_2(self):
        _input = [2, 1]
        self.sort(_input)
        self.assertSequenceEqual(_input, sorted(_input))

    def test_3(self):
        _input = [1]
        self.sort(_input)
        self.assertSequenceEqual(_input, sorted(_input))

    def test_4(self):
        _input = [3, 5, 4]
        self.sort(_input)
        self.assertSequenceEqual(_input, sorted(_input))

    def test_5(self):
        _input = [-1, 2, -6, 3, 5, 123, 54, 2, -6, 23, 4]
        self.sort(_input)
        self.assertSequenceEqual(_input, sorted(_input))

class SelectionSort(SortTest):
    '''
    Time Complexity: O(n^2)
    Space Complexity: O(1)
    '''
    def sort(self, nums):
        for i in range(len(nums) - 1):
            min_index = i
            for j in range(i + 1, len(nums)):
                min_index = min(j, min_index, key=lambda x: nums[x])
            nums[i], nums[min_index] = nums[min_index], nums[i]

class InsertionSort(SortTest):
    '''
    Time Complexity: O(n^2)
    Space Complexity: O(1)
    '''
    def sort(self, nums):
        for i in range(1, len(nums)):
            for j in range(i, 0, -1):
                if nums[j] < nums[j - 1]:
                    nums[j - 1], nums[j] = nums[j], nums[j - 1]
                else:
                    break

class BubbleSort(SortTest):
    '''
    Time Complexity: O(n^2)
    Space Complexity: O(1)
    '''
    def sort(self, nums):
        for i in range(len(nums) - 1):
            for j in range(len(nums) - i - 1):
                if nums[j] > nums[j + 1]:
                    nums[j], nums[j + 1] = nums[j + 1], nums[j]

class MergeSortTopDown(SortTest):
    def sort(self, nums):
        self.mergeSort(nums, 0, len(nums))

    def mergeSort(self, nums, lo, hi):
        if hi - lo > 1:
            mid = (lo + hi) // 2

            self.mergeSort(nums, lo, mid)
            self.mergeSort(nums, mid, hi)

            i, j = lo, mid
            arr = []
            while i < mid and j < hi:
                if nums[i] < nums[j]:
                    arr.append(nums[i])
                    i += 1
                else:
                    arr.append(nums[j])
                    j += 1
            while i < mid:
                arr.append(nums[i])
                i += 1
            while j < hi:
                arr.append(nums[j])
                j += 1

            nums[lo:hi] = arr

class MergeSortBottomUp(SortTest):
    def sort(self, nums):
        pass

class QuickSortLomuto(SortTest):
    def sort(self, nums):
        self.quickSort(nums, 0, len(nums) - 1)
        pass

    def quickSort(self, nums, lo, hi):
        if lo < hi:
            pivot = self.partition(nums, lo, hi)
            self.quickSort(nums, lo, pivot - 1)
            self.quickSort(nums, pivot + 1, hi)

    def partition(self, nums, lo, hi):
        i = lo
        for j in range(lo, hi):
            if nums[j] < nums[hi]:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1

        nums[i], nums[hi] = nums[hi], nums[i]

        return i

class QuickSortHoare(SortTest):
    def sort(self, nums):
        self.quickSort(nums, 0, len(nums) - 1)

    def quickSort(self, nums, lo, hi):
        if lo < hi:
            pivot = self.partition(nums, lo, hi)
            self.quickSort(nums, lo, pivot)
            self.quickSort(nums, pivot + 1, hi)

    def partition(self, nums, lo, hi):
        pivot = nums[(lo + hi) // 2]
        i = lo
        j = hi

        while True:
            while nums[i] < pivot:
                i += 1
            while nums[j] > pivot:
                j -= 1

            if i >= j:
                return j

            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1

class HeapSort(SortTest):
    def sort(self, nums):
        '''
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        '''
        self.heapify(nums)
        for end in range(len(nums) - 1, 0, -1):
            nums[0], nums[end] = nums[end], nums[0]
            self.siftDown(nums, 0, end)

    def heapify(self, nums):
        '''
        Time Complexity: O(n)
        Space Complexity: O(1)
        '''
        for start in range(len(nums) - 1, -1, -1):
            self.siftDown(nums, start, len(nums))
    
    def siftDown(self, nums, start, end):
        '''
        Time Complexity: O(log n)
        Space Complexity: O(1)
        '''
        root = swap = start
        left = 2*root + 1
        while left < end:
            if nums[swap] < nums[left]:
                swap = left
            if left + 1 < end and nums[swap] < nums[left + 1]:
                swap = left + 1
            
            if swap == root:
                return

            nums[root], nums[swap] = nums[swap], nums[root]
            root, left = swap, 2*swap + 1


def main():
    test_classes = [
        SelectionSort, 
        InsertionSort,
        BubbleSort,
        HeapSort, 
        MergeSortTopDown, 
        QuickSortLomuto, 
        QuickSortHoare
    ]
    tests = ['test_1', 'test_2', 'test_3', 'test_4', 'test_5']
    suite = unittest.TestSuite()
    for klass in test_classes:
        for test in tests:
            suite.addTest(klass(test))
    runner = unittest.TextTestRunner().run(suite)

if __name__ == '__main__':
    main()

File: algorithms/sort/sort.js
export function bubbleSort(arr) {
    for (let i = arr.length - 1; i >= 1; i--) {
        for (let j = 0; j < i; j++) {
            const lo = j;
            const hi = j + 1;
            if (arr[lo] > arr[hi]) {
                [arr[lo], arr[hi]] = [arr[hi], arr[lo]];
            }
        }
    }

    return arr;
}

export function insertionSort(arr) {
    for (let i = 1; i < arr.length; i++) {
        for (let j = i; j >= 1; j--) {
            const lo = j - 1;
            const hi = j;
            if (arr[lo] > arr[hi]) {
                [arr[lo], arr[hi]] = [arr[hi], arr[lo]];
            }
        }
    }

    return arr;
}

export function selectionSort(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        let minIndex = i;
        for (let j = i + 1; j < arr.length; j++) {
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        [arr[i], arr[minIndex]] = [arr[minIndex], arr[i]];
    }

    return arr;
}

export function mergeSortTopDown(arr) {
    const copyArr = Array(arr.length);
    for (let i = 0; i < arr.length; i++) {
        copyArr[i] = arr[i];
    }

    function merge(lo, hi, srcArr, destArr) {
        if (hi - lo <= 1) {
            return;
        }

        const mid = Math.floor((lo + hi) / 2);
        merge(lo, mid, destArr, srcArr);
        merge(mid, hi, destArr, srcArr);

        let i = lo;
        let j = mid;
        for (let k = lo; k < hi; k++) {
            if (j >= hi || (i < mid && srcArr[i] <= srcArr[j])) {
                destArr[k] = srcArr[i];
                i++;
            } else {
                destArr[k] = srcArr[j];
                j++;
            }
        }
    }
    merge(0, arr.length, copyArr, arr);

    return arr;
}

export function mergeSortBottomUp(arr) {
    const copyArr = Array(arr.length);
    for (let i = 0; i < arr.length; i++) {
        copyArr[i] = arr[i];
    }

    const log = Math.ceil(Math.log2(arr.length));
    let chunkLength = 2;
    let [srcArr, destArr] = log % 2 ? [copyArr, arr] : [arr, copyArr];
    for (let level = 0; level < log; level++) {
        const iterations = Math.ceil(arr.length / chunkLength);

        for (let iteration = 0; iteration < iterations; iteration++) {
            const lo = iteration * chunkLength;
            const hi = lo + chunkLength;
            const mid = Math.floor((lo + hi) / 2);
            const upper = Math.min(hi, arr.length);

            let i = lo;
            let j = mid;
            for (let k = lo; k < upper; k++) {
                if (j >= upper || (i < mid && srcArr[i] <= srcArr[j])) {
                    destArr[k] = srcArr[i];
                    i++;
                } else {
                    destArr[k] = srcArr[j];
                    j++;
                }
            }
            
        }

        chunkLength *= 2;
        [srcArr, destArr] = [destArr, srcArr];
    }

    return arr;
}

export function quickSortLomuto(arr) {
    function partition(lo, hi) {
        const pivotIndex = hi - 1;
        let partitionIndex = lo;

        for (let i = lo; i < pivotIndex; i++) {
            if (arr[i] <= arr[pivotIndex]) {
                [arr[i], arr[partitionIndex]] = [arr[partitionIndex], arr[i]];
                partitionIndex++;
            }
        }
        [arr[partitionIndex], arr[pivotIndex]] = [arr[pivotIndex], arr[partitionIndex]];

        return partitionIndex;
    }

    function quickSort(lo, hi) {
        if (hi - lo <= 1) {
            return;
        }

        const partitionIndex = partition(lo, hi);
        quickSort(lo, partitionIndex);
        quickSort(partitionIndex + 1, hi);
    }

    quickSort(0, arr.length);
    return arr;
}

export function quickSortHoare(arr) {
    function partition(lo, hi) {
        const pivot = arr[Math.floor((lo + hi) / 2)];
        let i = lo - 1;
        let j = hi;

        while (true) {
            do {
                i++;
            } while (arr[i] < pivot);
            do {
                j--;
            } while (arr[j] > pivot);

            if (i < j) {
                [arr[i], arr[j]] = [arr[j], arr[i]];
            } else {
                break;
            }
        }

        return i;
    }

    function quickSort(lo, hi) {
        if (hi - lo <= 1) {
            return;
        }

        const partitionIndex = partition(lo, hi);
        quickSort(lo, partitionIndex);
        quickSort(partitionIndex, hi);
    }

    quickSort(0, arr.length);
    return arr;
}

export function heapSort(arr) {
    function getLeftChildIndex(i) {
        return i * 2 + 1;
    }

    function getRightChildIndex(i) {
        return i * 2 + 2;
    }

    function siftDown(i, arr, length) {
        while (true) {
            const nodes = [i, getLeftChildIndex(i), getRightChildIndex(i)]
                .map(index => ({
                    index,
                    value: index < length ? arr[index] : -Infinity,
                }));
            const maxNode = nodes.reduce((maxNode, node) => node.value > maxNode.value ? node : maxNode);

            if (maxNode.index === i) {
                break;
            } else {
                [arr[i], arr[maxNode.index]] = [arr[maxNode.index], arr[i]];
                i = maxNode.index;
            }
        }
    }

    function heapify(arr) {
        for (let i = arr.length - 1; i >= 0; i--) {
            siftDown(i, arr, arr.length);
        }
    }

    function sort(arr) {
        for (let i = arr.length - 1; i >= 1; i--) {
            [arr[i], arr[0]] = [arr[0], arr[i]];
            siftDown(0, arr, i);
        }
    }

    heapify(arr);
    sort(arr);

    return arr;
}